### DISCUSSION TITLE
Custom Hook - useMutation

### DISCUSSION THREAD

**Discussion by jaydedm** [TanStack/query#1915] [Q&A]
Hey, everyone. This library is amazing! Thank you to all those involved with its development.

I was wondering if anyone could provide some guidance on best practices for my use case. 

I just want to extract my useMutations that are currently located inside my function component body to a helpers file.

It currently looks something like this:

```
const MyComponent = (): ReactElement => {
 
  const spPostResourceType = useMutation(
    async () => {
      return await sp.web.lists.getByTitle("Title").items.add({
        Title: newResourceTypeTitle,
        CountryId: selectedCountry,
      });
    },
    {
      onSuccess: () => {
        queryClient.invalidateQueries("myCache");
      },
    }
  );

return (
  <Button onClick={spPostResourcetype.mutate()}
)
}
```

Is it possible to move spPostResourceType out of this component and create a custom `usePostResourceType` hook without getting an invalid hook call outside of a function component?

Thanks!

**Comment by TkDodo** [comment]
> without getting an invalid hook call outside of a function component?

No, you can only call hooks from other hooks or fictional components. Those are the rules of hooks. Why would you need to do that?

**Reply by jaydedm** [reply]
I just don't like how bulky my component is getting. I was looking for ways to refactor and simplify the organization.

**Reply by TkDodo** [reply]
So, extract it to a custom hook? That has always been the recommendation

**Reply by jaydedm** [reply]
I understand, but I'm not doing it correctly. That's what I'm asking about! When I extract the code to a separate file I get an invalid hooks call error. I'm doing something wrong when creating a custom hook with `useMutation`. I have loads of custom hooks using `useQuery` that work just fine.

**Reply by TkDodo** [reply]
Did you name the function with `use`? Can you show your code?

**Reply by jaydedm** [reply]
absolutely!

```
const useDeleteResource = useMutation(
  async (selectedFilename) => {
    const domainPath =
      process.env.NODE_ENV === "dev" ? process.env.DOMAIN : "..";
    const url =
      domainPath +
      `/_api/web/lists/getbytitle('AC-HTA-Documents')/Files('${selectedFilename}')`;
    return await axiosDELETERequest(url);
  },
  {
    onSuccess: () => {
      const queryClient = useQueryClient();
      queryClient.invalidateQueries("resourcetype-resources");
    },
    onError: () => {
      const { addToast } = useToasts();
      addToast("You are unauthorized to make this action.", {
        appearance: "error",
        autoDismiss: true,
      });
    },
  }
);

export default useDeleteResource;
```

**Reply by jaydedm** [reply]
i still get invalid hook calls when commenting out the `onSuccess` and `onError` queryClient and addToast calls respectively.

**Reply by TkDodo** [reply]
You are calling useQueryClient _inside_ the onSuccess callback. That’s not allowed. Call it at the top of useDeleteResource

**Reply by jaydedm** [reply]
yes, even removing that I still get the same error.

**Reply by dusty** [reply]
Don't you need that hook to return a function?

```
const useDeleteResource = useMutation(...)
vs
const useDeleteResource = () => useMutation(...)
```

**Reply by jaydedm** [reply]
Good point @dusty.

Here's my current custom hook file then:

```
import { useMutation, useQueryClient } from "react-query";
import { useToasts } from "react-toast-notifications";
import { axiosDELETERequest } from "../serviceHelpers";

const queryClient = useQueryClient();
const { addToast } = useToasts();

const useDeleteResource = () => useMutation(
  async (selectedFilename) => {
    const domainPath =
      process.env.NODE_ENV === "dev" ? process.env.DOMAIN : "..";
    const url =
      domainPath +
      `/_api/web/lists/getbytitle('AC-HTA-Documents')/Files('${selectedFilename}')`;
    return await axiosDELETERequest(url);
  },
  {
    onSuccess: () => {
      queryClient.invalidateQueries("resourcetype-resources");
    },
    onError: () => {
      addToast("You are unauthorized to make this action.", {
        appearance: "error",
        autoDismiss: true,
      });
    },
  }
);

export default useDeleteResource;
```

And my component that uses the hook looks something like this:

```
const resourcesList = ():ReactElement => {
     const deleteResourceFile = useDeleteResource();
}
```

On app load, my app is crashing and giving me this error still:

```
 Invalid hook call. Hooks can only be called inside of the body of a function component
 ```

**Reply by jaydedm** [reply]
I can't use useQuery like that, right? Hmm..

**Reply by jaydedm** [reply]
changed my call to 

```
  const deleteResourceFile = useDeleteResource(queryClient, addToast, selectedFilename);
```

**Reply by jaydedm** [reply]
We're in business now. Thanks, boiz

**Reply by TkDodo** [reply]
You still can’t call useToast etc outside of a hook. It needs to be:

```
const useDeleteResource = () => {
    const queryClient = useQueryClient();
    const { addToast } = useToasts();

    return useMutation(...)
}
```
